### ISSUE TITLE
How can I enable NMM (Non Max Merge)?

### ISSUE THREAD

**Issue by Dasadaly** [ultralytics/ultralytics#22257]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I have one object highlighted with several masks; if you combine them, you get an excellent result. But I couldn't find any implementation of NMM. I need to combine all the results into one mask, for one object

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Dasadaly, thank you for your interest in Ultralytics 🚀! This is an automated response — an Ultralytics engineer will follow up soon to help further.

For general guidance, we recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples and where many of the most common questions may already be answered.

Since this is a segmentation-focused question about combining multiple masks for a single object (NMM-like behavior), please share a few details to help us assist:
- What task/model are you using (e.g., instance segmentation with YOLO11)?
- A short code snippet showing your inference pipeline
- Example images with the predicted masks and a brief note on the expected merged output

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a Python>=3.8 environment with PyTorch>=1.8 to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Thanks for the question—Ultralytics YOLO doesn’t include a built-in “Non‑Max Merge” for segmentation; NMS is applied in the segmentation postprocess but it suppresses rather than merges masks, see the segmentation postprocess reference and the non_max_suppression reference for details. You can merge overlapping masks after inference by grouping same‑class masks with high mask‑IoU and OR‑ing them, e.g.:

```python
from ultralytics import YOLO
import torch

model = YOLO("yolo11n-seg.pt")
r = model("image.jpg")[0]  # first image
m = r.masks.data.bool()    # (N, H, W)
cls, conf = r.boxes.cls, r.boxes.conf

thr = 0.5  # mask IoU threshold to merge
N = m.shape[0]
used = torch.zeros(N, dtype=torch.bool)
merged_masks, merged_cls, merged_conf = [], [], []

for i in range(N):
    if used[i]:
        continue
    group = [i]
    used[i] = True
    for j in range(i + 1, N):
        if used[j] or cls[j] != cls[i]:
            continue
        inter = (m[i] & m[j]).sum()
        union = (m[i] | m[j]).sum().clamp_min(1)
        if (inter.float() / union.float()) > thr:
            group.append(j)
            used[j] = True
    merged_masks.append(m[group].any(0))          # union
    merged_cls.append(cls[group][0])              # keep class
    merged_conf.append(conf[group].max())         # take max conf

merged_masks = torch.stack(merged_masks).float()  # your merged per-object masks
```

If you’re using SAM and only need non‑overlapping outputs (not merging), there’s a `non_overlap_masks` option in the SAM predictor’s postprocess. For context, see segmentation postprocess and NMS behavior in the docs: segmentation postprocess, non_max_suppression, and SAM postprocess non-overlapping masks.
